### ISSUE TITLE
Race condition in EventFileWriter

### ISSUE THREAD

**Issue by gibiansky** [tensorflow/tensorflow#6974]
There is a race condition here in `EventFileWriter`: [here](https://github.com/tensorflow/tensorflow/blob/287db3a9b0701021f302e7bb58af5cf89fdcd424/tensorflow/python/summary/writer/event_file_writer.py#L68)

```python
    if not gfile.IsDirectory(self._logdir):
      gfile.MakeDirs(self._logdir)
```

It is possible for multiple concurrent threads and/or processes to get to this code simultaneously, both check `IsDirectory`, return `False`, then both try creating the directory, and have one succeed and one error.

This can be fixed by by catching the `AlreadyExistsError` from `MakeDirs` and handling it gracefully.

(this is a real issue that we have encountered in daily usage and causes us issues, not just a hypothetical)

**Comment by asimshankar** [comment]
I don't believe it is recommended to have two distinct `EventFileWriters` writing to the same log directory as that could lead to two of them writing to the same file simultaneously, which will certainly be incorrect.

Could you elaborate on how you run into this situation and why you need to have concurrent `EventFileWriters` in the same log directory? Can you have your concurrent writers writing to different directories?

**Comment by gibiansky** [comment]
I agree that this can be handled more effectively by fixing user code, but it is still an issue that should be fixed and should be causing at most a warning (not a crash).

The reason this is happening is because we are training in parallel using MPI, and so we have many copies of identical MPI processes that run the same model (but communicate with each other to reduce gradients). If the processes are identical they will each write a log of the things they are doing to each directory. This _could_ be handled by either doing a per-rank log directory or by having only a single rank write logs, and that is how we will handle this for the time being.

**Comment by yaroslavvb** [comment]
We are also experimenting with MPI for training, seems like catching/ignoring `AlreadyExistsError ` is a an easy fix.

**Comment by asimshankar** [comment]
It won't be as simple as that, since we'd want to ensure that the underlying [C++ `EventsWriter`](https://github.com/tensorflow/tensorflow/blob/754048a0453a04a761e112ae5d99c149eb9910dd/tensorflow/core/util/events_writer.h#L43) doesn't end up with the same filename and that we are not writing concurrently to the same underlying file.

But that said, it shouldn't be too hard to fix the whole chain. Contributions welcome!

**Comment by gibiansky** [comment]
These are somewhat orthogonal issues. I can trigger this issue without doing multiple training jobs at once – for example, I can have a loop that runs a small TensorFlow script with logging over and over again and writing to dir `log-dir`, and concurrently another process that is constantly looping and doing `rm -rf log-dir`. Eventually you'll hit this race condition. 

Do you think a fix for this issue could be merged _without_ having to drag in coordination between multiple processes doing training? That may be our specific use case, but it has nothing to do with what's causing the bug...

**Comment by gibiansky** [comment]
In particular, the using-the-same-filename issue is _already_ present: suppose the directory already exists in advance, and you run two jobs training and logging to the same directory. They might write to the same file already as it is now – changing the semantics here does not affect that bug. Since these are two separate bugs they can be fixed in two separate issues and PRs.

**Comment by asimshankar** [comment]
I was thinking about things working end-to-end, but as you correctly point out, there is no strict dependency. 

That said, I'm not sure if the right fix is to catch and ignore the `AlreadyExistsError` or if there should be no error in the first place. For example, the underlying C++ operation should _not_ be failing (https://github.com/tensorflow/tensorflow/blob/5e5dc97dd3523509ce5f536d7be5122d016fc6b5/tensorflow/core/platform/file_system.h#L107) 

So I'd like to understand why the error is being thrown. Doesn't happen in this simple test:
```python
import tensorflow as tf
tf.platform.gfile.MakeDirs('/tmp/foo')
tf.platform.gfile.MakeDirs('/tmp/foo')
```

Let me dig around a bit as to why you're getting this error, or if you have any pointers, that would be help.

**Comment by asimshankar** [comment]
It seems there is some inconsistency between implementations of `FileSystem::CreateDir`. For example, [posix_file_system.cc:230](https://github.com/tensorflow/tensorflow/blob/5e5dc97dd3523509ce5f536d7be5122d016fc6b5/tensorflow/core/platform/posix/posix_file_system.cc#L230) fails when the directory already exist, while the [test filesystem](https://github.com/tensorflow/tensorflow/blob/ec7929b878926c39255254e9aea992f0bc65aa68/tensorflow/core/platform/file_system_test.cc#L48) does not.

I'll look into making these `FileSystem` implementations consistent.

**Comment by gibiansky** [comment]
Thanks!
